homeassistant/auth/__init__.py

```diff
@@ -474,10 +485,22 @@ async def async_validate_access_token(
             issuer = refresh_token.id
 
         try:
-            jwt.decode(token, jwt_key, leeway=10, issuer=issuer, algorithms=["HS256"])
+            claims = jwt.decode(
+                token, jwt_key, leeway=10, issuer=issuer, algorithms=["HS256"]
+            )
         except jwt.InvalidTokenError:
             return None
 
+        hass_claims = claims.get(CLAIM_HASS_NAMESPACE, {})
+        remote_claim = hass_claims.get(CLAIM_HASS_REMOTE)
+        provider_claim = hass_claims.get(CLAIM_HASS_PROVIDER)
+        if (
+            remote_claim
+            and provider_claim == "trusted_networks"
+            and remote != remote_claim
+        ):
+            return None
```

Comment:
This check should be done in the auth provider itself rather than here. Ie we should provide the auth provider with the set of hass claims, and let it decide if the access token i valid.

This code path should be quick thou. So I already think it is wrong to lookup the refresh token here. But i suppose for long lived access tokens it's the only way :(

Reply:
we could completely handle this by the known refresh token at this stage i suppose if we keep the code around that looks up the refresh_token here.

Reply:
I've dropped the method of using claims in the access token since we already know the refresh token at this stage. I would have preferred to use claims here to avoid hitting the auth backend. But code paths would have ended up different for the refresh token step.